src/models/base_model.py

```diff
@@ -0,0 +1,218 @@
+
+    This class supports adding feature hooks to some model layers.
+    Class has feature info list for every hook.
+    To create hooks, method self.get_features_info() must be rewrited!
+    Also class contains method get_output_channels wich returns forward pass and hooks output channels. 
+    """
+    def __init__(self):
+        """Inits BaseModel class and it's hooks.
+
+        Raises:
+            ValueError: If rewrited self.get_features_info() method return not FeatureInfo list.
+        """
+        super().__init__()
+
+        # Create feature_info list with hooks
+        self._features_info, self._feature_hooks = self._create_hooks()
+
+
+    def get_features_info(self) -> List[FeatureInfo]:
+        """Method for initialize Hooks.
+        
+        It must be rewrite in inherited classes, if Module would need hooks.
+
+        Retruns: Hooks FeatureInfo list.
+        """
+        return None
+
+    def _create_hooks(self) -> Tuple[List[FeatureInfo], FeatureHooks]:
+        """Call self.get_features_info() method to generate feature_info list and then generate feature hooks.
+        
+        If self.get_features_info() not rewrited in inheritable class the features_info and hooks would be None.
+
+        Returns:
+            feature_info: Hooks feature info list.
+            feature_hooks: Generated hooks.
+        """
+        features_info = self.get_features_info()
+        # Check if all self._feature_info is FeatureInfo types.
+        if features_info is not None:
+            self.__check_features_info_types()
+            hooks = [
+                Hook(module_name=feature.module_name, hook_type=feature.hook_type) 
+                for feature in features_info
+                ]
+            feature_hooks = FeatureHooks(hooks, self.named_modules())
+        else:
+            feature_hooks = None
+
+        return features_info, feature_hooks
+
+    def __check_features_info_types(self):
+        """Check if all self._feature_info is FeatureInfo class.
+            
+        Raises:
+            ValueError: If any value in feature_info is not FeatureInfo class.
+        """
+        for feature in self._features_info:
+            if type(feature) != FeatureInfo:
+                raise ValueError('All feature_info must be FeatureInfo class.')
+
+    def forward(self, x: torch.Tensor):
+        # Return features for classification.
+        raise NotImplementedError
+
+    def forward_stage_features(self, x: torch.Tensor) -> Tuple[torch.Tensor, List[torch.Tensor]]:
+        """Return model output with hooks features.
+        
+        Returns:
+            last_features: Module forward method output.
+            hooks_features: Hooks outputs.
+        """
+        last_features = self.forward(x)
+        hooks_features = self._feature_hooks.get_features(x.device)
+        hooks_features = list(hooks_features.values())
+        hooks_features = [x] + hooks_features
```

Comment:
I don't think you should add `x` to list of outputs. Since the user passed `x` as input he already has this value